### ISSUE TITLE
Stateless test `02784_parallel_replicas_automatic_decision_join` is flaky 

### ISSUE THREAD

**Issue by SmitaRKulkarni** [ClickHouse/ClickHouse#60625]
Sample fail : 
https://s3.amazonaws.com/clickhouse-test-reports/60251/bb7453a339daa023b6613dc18788e63cf36adff3/stateless_tests__asan__[2_4].html

Recent fails : [Link](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSBjaGVja19uYW1lIExJS0UgJ1N0YXRlJScKICAgIEFORCBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gSU5URVJWQUwgMTAwIEhPVVIKICAgIEFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gMAogICAgQU5EIHRlc3Rfc3RhdHVzICE9ICdTS0lQUEVEJwogICAgQU5EICh0ZXN0X3N0YXR1cyBMSUtFICdGJScgT1IgdGVzdF9zdGF0dXMgTElLRSAnJUUlJykKICAgIEFORCBjaGVja19zdGF0dXMgIT0gJ3N1Y2Nlc3MnCiAgICBBTkQgdGVzdF9uYW1lIExJS0UgJyUwMjc4NF9wYXJhbGxlbF9yZXBsaWNhc19hdXRvbWF0aWNfZGVjaXNpb25fam9pbiUnCk9SREVSIEJZIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgY2hlY2tfc3RhcnRfdGltZQ==) 

`2024.02.29 16:13:04.520866 [ 1804 ] {a2811e0c-a3ad-4bdd-86bc-03392389dec9} <Error> executeQuery: Code: 210. DB::NetException: Connection reset by peer, while writing to socket ([::ffff:127.0.0.2]:9000 -> [::ffff:127.0.0.1]:43670). (NETWORK_ERROR) (version 24.2.1.1) (from [::ffff:127.0.0.1]:43670) (comment: 02784_parallel_replicas_automatic_decision_join.sh) (in query: SELECT sum(`value`) FROM `test_kzmkij6r`.`test_parallel_replicas_automatic_left_side` ALL INNER JOIN `test_parallel_replicas_automatic_count_right_side` AS `test_parallel_replicas_automatic_count_right_side` USING (`number`)), Stack trace (when copying this message, always include the lines below):`

**Comment by Algunenano** [comment]
There is something odd going on. The test executes locally in ~1.5 seconds.

Looking at the logs I see one of the queries expending 150+ seconds aggregating, which is madness:
```
2024.03.01 05:30:22.831461 [ 2470 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregating
2024.03.01 05:30:22.831690 [ 2470 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> Aggregator: Aggregation method: without_key
2024.03.01 05:30:23.606207 [ 5778 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 196481 to 1 rows (from 3.00 MiB) in 5.275584305 sec. (37243.458 rows/sec., 581.93 KiB/sec.)
2024.03.01 05:30:24.047542 [ 5787 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 65536 to 1 rows (from 1.00 MiB) in 5.71673505 sec. (11463.886 rows/sec., 179.12 KiB/sec.)
2024.03.01 05:30:24.088826 [ 8249 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.757962629 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.113664 [ 5785 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.783115492 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.114678 [ 5782 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.784193427 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.117171 [ 5770 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.786516194 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.128342 [ 8239 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 65409 to 1 rows (from 1022.02 KiB) in 5.79776551 sec. (11281.760 rows/sec., 176.28 KiB/sec.)
2024.03.01 05:30:24.153829 [ 8243 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.823449669 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.171444 [ 3628 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.840931782 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.175538 [ 2470 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 127 to 1 rows (from 1.98 KiB) in 5.844865101 sec. (21.728 rows/sec., 347.66 B/sec.)
2024.03.01 05:30:24.190088 [ 3630 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.859281488 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.192556 [ 1646 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.861931988 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.226842 [ 5794 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 5.896443708 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:24.243535 [ 8323 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 172447 to 1 rows (from 2.63 MiB) in 5.913034498 sec. (29163.875 rows/sec., 455.69 KiB/sec.)
2024.03.01 05:30:24.429238 [ 5792 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 0 to 1 rows (from 0.00 B) in 6.098470613 sec. (0.000 rows/sec., 0.00 B/sec.)
2024.03.01 05:30:41.078570 [ 8241 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregating
2024.03.01 05:30:41.079016 [ 8241 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> Aggregator: Aggregation method: without_key
2024.03.01 05:33:08.269268 [ 8241 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> AggregatingTransform: Aggregated. 500000 to 1 rows (from 11.44 MiB) in 169.938295761 sec. (2942.244 rows/sec., 68.96 KiB/sec.)
2024.03.01 05:33:08.269528 [ 8241 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> Aggregator: Merging aggregated data
2024.03.01 05:33:08.269947 [ 8241 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> Aggregator: Statistics updated for key=2645075937495578225: new sum_of_sizes=16, median_size=1
2024.03.01 05:33:08.284266 [ 8827 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Debug> executeQuery: Read 12000000 rows, 118.44 MiB in 169.983156 sec., 70595.230035616 rows/sec., 713.51 KiB/sec.
2024.03.01 05:33:08.296349 [ 8827 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Trace> HashJoin: Join data is being destroyed, 144965632 bytes and 1000000 rows in hash table
2024.03.01 05:33:08.348865 [ 8827 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Debug> MemoryTracker: Peak memory usage (for query): 255.47 MiB.
2024.03.01 05:33:08.349117 [ 8827 ] {02784_automatic_parallel_replicas_join-test_inzv6qet_simple_join_10M_pure} <Debug> TCPHandler: Processed in 170.050514961 sec.
```

**Comment by Algunenano** [comment]
Looked into some other slow runs and I see the same, some aggregations taking way too long:

```
2024.03.05 09:49:36.728655 [ 21608 ] {78ce93d5-4382-456c-9c29-f3aaa4d05796} <Trace> AggregatingTransform: Aggregating
2024.03.05 09:49:36.728760 [ 21608 ] {78ce93d5-4382-456c-9c29-f3aaa4d05796} <Trace> Aggregator: An entry for key=7784547741159051238 found in cache: sum_of_sizes=31, median_size=1
2024.03.05 09:49:36.728836 [ 21608 ] {78ce93d5-4382-456c-9c29-f3aaa4d05796} <Trace> Aggregator: Aggregation method: without_key
2024.03.05 09:51:47.529849 [ 21608 ] {78ce93d5-4382-456c-9c29-f3aaa4d05796} <Trace> AggregatingTransform: Aggregated. 458752 to 1 rows (from 10.50 MiB) in 158.776095443 sec. (2889.301 rows/sec., 67.72 KiB/sec.)
2024.03.05 09:51:47.529944 [ 21608 ] {78ce93d5-4382-456c-9c29-f3aaa4d05796} <Trace> Aggregator: Merging aggregated data
```

Trying to understand why it's so slow before I reduce the table sizes in the tests

**Comment by Algunenano** [comment]
Trying to understand what happened to make this test so slow now:

```
WITH
    60 AS last_days_to_check,
    100 AS max_tests_to_show,
    5 as max_prs_to_show,
    70 AS test_name_max_length
SELECT
  toStartOfDay(check_start_time) as day, count(), arraySlice(groupArrayDistinct(pull_request_number), 1, 5) AS prs
FROM checks
WHERE
    check_start_time >= now() - interval last_days_to_check day
    AND check_status != 'success'
    AND ((check_name LIKE 'AST%' AND test_status = '') OR (test_status not in ('OK', 'SKIPPED', '', 'BROKEN', 'NOT_FAILED')))
    AND check_name not like 'Performance%' and check_name NOT IN ('Style check', 'Docs Check')
    AND test_name = '02784_parallel_replicas_automatic_decision_join'
    AND test_duration_ms > 500000
GROUP BY day
ORDER BY day DESC
LIMIT max_tests_to_show

```

Results:


№ | day | count() | prs | first_report_of_the_day
-- | -- | --
1 | 2024-03-05 00:00:00 | 29 | [0,60514,60615,60824,60770]
2 | 2024-03-04 00:00:00 | 20 | [0,60749,60308,60769,60765]
3 | 2024-03-03 00:00:00 | 14 | [0,60421,60468,60689,60694]
4 | 2024-03-02 00:00:00 | 3 | [60165,60518,58845]
5 | 2024-03-01 00:00:00 | 19 | [0,60657,60377,59929,60523]
6 | 2024-02-29 00:00:00 | 18 | [0,60383,60251,60348,59965]
7 | 2024-02-28 00:00:00 | 1 | [0]
8 | 2024-02-23 00:00:00 | 1 | [60338]
9 | 2024-02-18 00:00:00 | 2 | [60050]
10 | 2024-02-17 00:00:00 | 1 | [60050]
11 | 2024-02-16 00:00:00 | 1 | [60050]
12 | 2024-02-04 00:00:00 | 1 | [58845]
13 | 2024-01-15 00:00:00 | 2 | [55656]
14 | 2024-01-11 00:00:00 | 1 | [58717]
15 | 2024-01-08 00:00:00 | 1 | [0]

The first finger seems to point towards https://github.com/ClickHouse/ClickHouse/pull/60050 as before it was merged there were occasional timeouts, but now they happen 20+ times a day. It needs confirmation. cc @alexey-milovidov in case it rings any bells

**Comment by Algunenano** [comment]
Another alleyway I'm exploring is contention of `HashTablesStatistics` which has double mutex and no metrics about lock time

**Comment by nickitat** [comment]
after filtering out the noise (runs when 10+ tests failed because of the dead server or Igor's pr-s with refactorings around PRs) i found only one recent (after we disabled the test under sanitizers) and relevant failure: https://s3.amazonaws.com/clickhouse-test-reports/65705/89a8925f80c1174ce8546758cc8155cb53805b00/fast_test.html
it is addressed in #66905
